### ISSUE TITLE
1.4.0b2+ no longer calls dispose on InstrumentationManager

### ISSUE THREAD

**Issue by icemac** [sqlalchemy/sqlalchemy#6390]
**Describe the bug**
I have a custom `InstrumentationManager` extending `sqlalchemy.ext.instrumentation.InstrumentationManager`.

`InstrumentationManager` uses `_default_class_manager` as the attribute where it stores the class manager instead of the default one `_sa_class_manager`.

**Expected behavior**
`sqlalchemy.orm.clear_mappers()` should remove the attribute `_default_class_manager` from the mapped class. So creating a new mapper afterwards is possible.

Until https://github.com/sqlalchemy/sqlalchemy/commit/5ec5b0a6c7b618bba7926e21f77be9557973860f#diff-5e3090555a0af448d2716136198d1b606fcbc5253df1065c04b4abaae1ac2e60L541 this was the case because `manager.dispose` delegated to `InstrumentationManager.dispose` which removes the attribute.

With the change in 1.4.0b2 `ClassManager` just tries to remove the attribute named `self.MANAGER_ATTR` (which in this case has the value of `'_sa_class_manager'`).
See https://github.com/sqlalchemy/sqlalchemy/commit/5ec5b0a6c7b618bba7926e21f77be9557973860f#diff-5e3090555a0af448d2716136198d1b606fcbc5253df1065c04b4abaae1ac2e60R357

**To Reproduce**

```py
from sqlalchemy.ext import instrumentation
from sqlalchemy import Column, Integer, Table, String, create_engine
import sqlalchemy.orm
import sqlalchemy.schema


metadata = sqlalchemy.schema.MetaData()
metadata.bind = create_engine('sqlite://')

table = Table(
    'my-table', metadata,
    Column('mytable_id', Integer, primary_key=True),
    Column('value', String(50)))
metadata.create_all()


class MyInstrumentation(instrumentation.InstrumentationManager):
    """The actual implementation does not seem to matter."""

    def dispose(self, class_):
        # workaround required for SQLAlchemy 1.3:
        # instrumentation.InstrumentationManager.dispose()
        # needs 3 arguments, but is called with only 2 arguments.
        # (Not called in SQLAlchemy 1.4)
        return super().dispose(class_, manager=None)

    # My current workaround for SQLAlchemy 1.4.x:
    # def manage(self, class_, manager):
    #     manager.MANAGER_ATTR = '_default_class_manager'
    #     super().manage(class_, manager)


instrumentation.instrumentation_finders.append(
    lambda cls: MyInstrumentation)


def extended_manager_of_class(cls):
    """Redirect to the attribute where the manager class is stored."""
    return cls.__dict__.get('_default_class_manager', None)

instrumentation._instrumentation_factory.manager_of_class = (
    extended_manager_of_class)


class MyTable:
    mytable_id = None
    value = None

mapper = sqlalchemy.orm.mapper(MyTable, table)

sqlalchemy.orm.clear_mappers()

mapper = sqlalchemy.orm.mapper(MyTable, table)

print('Success!')
```

**Error**

happens on the second call to `sqlalchemy.orm.mapper`: 
```
Traceback (most recent call last):
  File "example.py", line 53, in <module>
    mapper = sqlalchemy.orm.mapper(MyTable, table)
  File "<string>", line 2, in mapper
  File "<string>", line 2, in __init__
  File ".../sqlalchemy/util/deprecations.py", line 298, in warned
    return fn(*args, **kwargs)
  File ".../sqlalchemy/orm/mapper.py", line 683, in __init__
    self._configure_class_instrumentation()
  File ".../sqlalchemy/orm/mapper.py", line 1214, in _configure_class_instrumentation
    % self.class_
sqlalchemy.exc.ArgumentError: Class '<class '__main__.MyTable'>' already has a primary mapper defined.
```

**Versions.**
 - OS: MacOS 10.15.7
 - Python: 3.7.10
 - SQLAlchemy: 1.4.11
 - Database: any (not dependent on database)

**Additional context**

* Running the same example on the latest SQLAlchemy 1.3 prints 'Success' as there is no exception.
* I added my current workaround as a comment to the example. – Maybe this is not the way to go but it seems to work.

**Have a nice day!** 😀

**Comment by zzzeek** [comment]
it looks like dispose is not working for any registry right now, the included example doesn't work for dispose:

```diff
diff --git a/examples/custom_attributes/custom_management.py b/examples/custom_attributes/custom_management.py
index 6cddfe7bd1..3613205c0f 100644
--- a/examples/custom_attributes/custom_management.py
+++ b/examples/custom_attributes/custom_management.py
@@ -17,7 +17,7 @@ from sqlalchemy import MetaData
 from sqlalchemy import Table
 from sqlalchemy import Text
 from sqlalchemy.ext.instrumentation import InstrumentationManager
-from sqlalchemy.orm import mapper
+from sqlalchemy.orm import clear_mappers, mapper
 from sqlalchemy.orm import relationship
 from sqlalchemy.orm import Session
 from sqlalchemy.orm.attributes import del_attribute
@@ -122,3 +122,5 @@ if __name__ == "__main__":
 
     a1 = sess.query(A).get(a1.id)
     assert len(a1.bs) == 1
+
+    clear_mappers()
```


```
   mapperlib._dispose_registries(mapperlib._all_registries(), False)
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/mapper.py", line 3462, in _dispose_registries
    reg._dispose_manager_and_mapper(manager)
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/decl_api.py", line 728, in _dispose_manager_and_mapper
    instrumentation._instrumentation_factory.unregister(class_)
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/ext/instrumentation.py", line 162, in unregister
    super(ExtendedInstrumentationRegistry, self).unregister(class_)
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/instrumentation.py", line 540, in unregister
    manager.unregister()
AttributeError: 'NoneType' object has no attribute 'unregister'
```

I'm not sure about your use of `instrumentation._instrumentation_factory.manager_of_class` which seems to sidestep that part, ideally you would not have to access any private variables to accomplish what you need.

**Comment by zzzeek** [comment]
possible fix for that last part

```diff
diff --git a/lib/sqlalchemy/ext/instrumentation.py b/lib/sqlalchemy/ext/instrumentation.py
index 300d0d81cf..b04fb57f04 100644
--- a/lib/sqlalchemy/ext/instrumentation.py
+++ b/lib/sqlalchemy/ext/instrumentation.py
@@ -155,11 +155,11 @@ class ExtendedInstrumentationRegistry(InstrumentationFactory):
         return factories
 
     def unregister(self, class_):
+        super(ExtendedInstrumentationRegistry, self).unregister(class_)
         if class_ in self._manager_finders:
             del self._manager_finders[class_]
             del self._state_finders[class_]
             del self._dict_finders[class_]
-        super(ExtendedInstrumentationRegistry, self).unregister(class_)
 
     def manager_of_class(self, cls):
         if cls is None:
```

I dont' have a lot of time to work on this area which TBH I would not have imagined anyone is using, I will have to find some time to get to it.  if you have fixes on your end with test cases to contribute, that would be helpful.

**Comment by icemac** [comment]
@zzzeek Thank you for looking into the issue. TBH I have the same impression that a custom instrumentation manager is not very widely used.
I will have a look into your suggestions in the next weeks (as I found a workaround the pressure for the customer is not that high, but I hope to get the time eventually.)

**Comment by zzzeek** [comment]
same here, I did do a refactoring of instrumentation for 1.4 and clearly this case was not tested so this is ultimately on my part to get it fixed, it's just lower priority than all the other things I broke in 1.4.

**Comment by zzzeek** [comment]
the restoration here is just to put it back:

```diff
diff --git a/lib/sqlalchemy/orm/instrumentation.py b/lib/sqlalchemy/orm/instrumentation.py
index 1edcab72a8..68ac3c2293 100644
--- a/lib/sqlalchemy/orm/instrumentation.py
+++ b/lib/sqlalchemy/orm/instrumentation.py
@@ -536,6 +536,7 @@ class InstrumentationFactory(object):
     def unregister(self, class_):
         manager = manager_of_class(class_)
         manager.unregister()
+        manager.dispose()
         self.dispatch.class_uninstrument(class_)
 
 

```

the other issue, adding `clear_mappers()` to the custom_attributes example, fails on 1.3 as well, same error: "NoneType' object has no attribute 'unregister'"

this bit in the example given above seems to be a workaround for this:

```python
instrumentation._instrumentation_factory.manager_of_class = (
    extended_manager_of_class)

```


but that's private API and the dispose process here should not lose the manager finders early.     to fix that, which would fix 1.3 also if we wanted, looks like this:

```diff
diff --git a/lib/sqlalchemy/ext/instrumentation.py b/lib/sqlalchemy/ext/instrumentation.py
index 300d0d81cf..b04fb57f04 100644
--- a/lib/sqlalchemy/ext/instrumentation.py
+++ b/lib/sqlalchemy/ext/instrumentation.py
@@ -155,11 +155,11 @@ class ExtendedInstrumentationRegistry(InstrumentationFactory):
         return factories
 
     def unregister(self, class_):
+        super(ExtendedInstrumentationRegistry, self).unregister(class_)
         if class_ in self._manager_finders:
             del self._manager_finders[class_]
             del self._state_finders[class_]
             del self._dict_finders[class_]
-        super(ExtendedInstrumentationRegistry, self).unregister(class_)
 
     def manager_of_class(self, cls):
         if cls is None:
```

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **master** branch:

**Ensure extended instrumentation is fully disposed** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2832
